packages/vats/src/lib-chainPublish.js

```diff
@@ -0,0 +1,46 @@
+// @ts-check
+
+import { E } from '@endo/far';
+
+/**
+ * Publish results of an async iterable into a chain storage node as an array
+ * serialized using JSON.stringify or an optionally provided function (e.g.,
+ * leveraging a serialize function from makeMarshal).
+ *
+ * @param {AsyncIterator} iterable
+ * @param {ReturnType<typeof import('./lib-chainStorage.js').makeChainStorageRoot>} chainStorageNode
+ * @param {{ timerService: ERef<TimerService>, serialize?: (obj: any) => string }} powers
+ */
+export async function publishToChainNode(
+  iterable,
+  chainStorageNode,
+  { timerService, serialize = JSON.stringify },
+) {
+  let nextIndex = 0n;
+  let isNewBlock = true;
+  let oldResults = [];
+  let results = [];
+  for await (const result of iterable) {
+    if (isNewBlock) {
+      isNewBlock = false;
+      oldResults = results;
+      results = [];
+      E(timerService)
+        .delay(1n)
+        .then(() => {
+          isNewBlock = true;
+        });
```

Comment:
Is this the right/best way to detect a new block?

Reply:
There's precedent for it. I'm not sure that it's guaranteed to detect a new block, but I think it's pretty close in practice. Does anything depend on this happening exactly once per block?

Reply:
> Does anything depend on this happening exactly once per block?

@dckc No, but it would be a problem for this to happen _more_ than once per block because then results produced earlier in the block could be lost.

> this functionality belongs in Golang

@michaelfig How exactly would that work, given that many vats will have chain publishers that need to purge old results? I guess we could create a new vat that somehow gets end-of-block messages and is the sole source of subscriptions that publish to chain storage, but a) I'd need help with the first part, and b) I'm not sure using such a vat would actually be an improvement over using `timerService` (although I'm willing to believe it might be—it would really help to see the code that actually needs chainStorage-backed subscriptions rather than guessing from the bottom up like this).